### DISCUSSION TITLE
Strongswan VPN ping issue

### DISCUSSION THREAD

**Discussion by SB8767** [strongswan/strongswan#2272] [Q&A]
I'm facing an issue with a client that I set up a StrongSwan VPN with. The client cannot ping my server when the tunnel is up. Also, the client has pings disabled on their subnet.

My server has 2 VPNs running (one setup with IKEv1 which works fine and the other VPN which I'm currently facing an issue).

See below ifconfig setup:

eth0:1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.108.40  netmask 255.255.255.248  broadcast 192.168.108.47
        ether 16:a7:4d:68:fa:2f  txqueuelen 1000  (Ethernet)

VPN tunnel is up:
`sudo ipsec status`
Security Associations (2 up, 0 connecting):
     vpn-deleted[52]: ESTABLISHED 77 seconds ago

$ sudo ip route show
192.168.108.40/29 dev eth0 proto kernel scope link src 192.168.108.40 

`iptables -L output`:
ACCEPT     all  --  10.240.209.128/29    admin-2/29           policy match dir in pol ipsec reqid 8 proto esp
ACCEPT     all  --  admin-2/29           10.240.209.128/29    policy match dir out pol ipsec reqid 8 proto esp
ACCEPT     icmp --  anywhere             anywhere             icmp echo-request
The CX is pinging from 10.240.209.129 to 192.168.108.41 and getting no response. 

`tcpdump -i eth0` shows me my home IP ssh'ing on to the server.

conn vpn-test
    type=tunnel
    auto=start
    keyexchange=ikev2
    authby=secret
    left=*serverpublicp*
    leftsubnet=192.168.108.40/29
    leftid=10.131.112.99
    leftfirewall=yes
    right=*clientpublicip*
    rightsubnet=10.240.209.128/29
    rightfirewall=yes
    ike=aes256-sha512-modp2048,aes256-sha384-modp2048,aes256-sha256-modp2048,aes256-sha-modp2048!
    esp=aes256-sha512-modp2048,aes256-sha384-modp2048,aes256-sha256-modp2048,aes256-sha1-modp2048!
    dpdaction=restart
    ikelifetime=86400s
    lifetime=3600s
    dpddelay=20s
    dpdtimeout=60s

What has been missed in the config/setup to cause the remote host not to be able to ping my eth0:1 subnet?

**Comment by tobiasbrunner** [comment]
> The client cannot ping my server when the tunnel is up.

What kind of client?

> Also, the client has pings disabled on their subnet.

What does that mean?

> VPN tunnel is up:
> `sudo ipsec status`
> Security Associations (2 up, 0 connecting):
> vpn-deleted[52]: ESTABLISHED 77 seconds ago

Please post the complete output (preferably of `ipsec statusall`) because this does not show whether there is a CHILD_SA established.

> $ sudo ip route show

strongSwan usually installs routes in table 220 for established CHILD_SAs. If not, check the log.



> What has been missed in the config/setup to cause the remote host not to be able to ping my eth0:1 subnet?

`eth0:1` is a legacy thing (as is ifconfig). This is just an additional address on eth0 with that label (check `ip addr`). Note that `192.168.108.40` is technically the network ID of that `/29` subnet, the first assignable address is `192.168.108.41`.

**Reply by SB8767** [reply]
> What kind of client?
>I believe they have a Cisco ASA
 
> What does that mean?
> From my end, I cannot ping their 10.240.209.128/29 subnet.
 
> Please post the complete output (preferably of `ipsec statusall`) because this does not show whether there is a CHILD_SA established.
> Security Associations (2 up, 0 connecting):
     vpn-csl[53]: ESTABLISHED 18 minutes ago, myip [10.131.112.99]...customerip [customerip]
     vpn-csl[53]: IKEv2 SPIs: 96b8d8a9ed6f45e6_i* a3dbc4c5f7b76d49_r, pre-shared key reauthentication in 23 hours
     vpn-csl[53]: IKE proposal: AES_CBC_256/HMAC_SHA2_512_256/PRF_HMAC_SHA2_512/MODP_2048
     vpn-csl{23}:  INSTALLED, TUNNEL, reqid 9, ESP SPIs: cfa6a010_i e05fdc27_o
     vpn-csl{23}:  AES_CBC_256/HMAC_SHA2_512_256, 0 bytes_i, 0 bytes_o, rekeying in 25 minutes
     vpn-csl{23}:   192.168.108.40/29 === 10.240.209.128/29
     
> > $ sudo ip route show
> 
> strongSwan usually installs routes in table 220 for established CHILD_SAs. If not, check the log.
>  there is nothing in /etc/strongswan but it is installed, which is odd.

```
sudo ipsec --version
Linux strongSwan U5.8.2/K5.4.0-170-generic
```

> `eth0:1` is a legacy thing (as is ifconfig). This is just an additional address on eth0 with that label (check `ip addr`). Note that `192.168.108.40` is technically the network ID of that `/29` subnet, the first assignable address is `192.168.108.41`.
>
```
 ip addr return: 
  inet 192.168.108.40/29 brd 192.168.108.47 scope global eth0:1
       valid_lft forever preferred_lft forever
```

**Reply by SB8767** [reply]
Sorry,  I found the logs:


847345:Jun  3 10:23:34 admin-2 charon: 13[NET] sending packet: from myip[500] to customerip[500] (596 bytes)
847346:Jun  3 10:23:34 admin-2 charon: 06[NET] received packet: from customerip [500] to myip[500] (619 bytes)
847347:Jun  3 10:23:34 admin-2 charon: 06[ENC] parsed IKE_SA_INIT response 0 [ SA KE No V V N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) V ]
847348:Jun  3 10:23:34 admin-2 charon: 06[IKE] received Cisco Delete Reason vendor ID
847349:Jun  3 10:23:34 admin-2 charon: 06[IKE] received Cisco Copyright (c) 2009 vendor ID
847350:Jun  3 10:23:34 admin-2 charon: 06[IKE] received FRAGMENTATION vendor ID
847351:Jun  3 10:23:34 admin-2 charon: 06[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_512_256/PRF_HMAC_SHA2_512/MODP_2048
847352:Jun  3 10:23:34 admin-2 charon: 06[IKE] received 2 cert requests for an unknown ca
847353:Jun  3 10:23:34 admin-2 charon: 06[IKE] authentication of '10.131.112.99' (myself) with pre-shared key
847354:Jun  3 10:23:34 admin-2 charon: 06[IKE] establishing CHILD_SA vpn-csl{24}
847355:Jun  3 10:23:34 admin-2 charon: 06[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
847356:Jun  3 10:23:34 admin-2 charon: 06[NET] sending packet: from myip[4500] to  customerip [4500] (480 bytes)
847357:Jun  3 10:23:34 admin-2 charon: 09[NET] received packet: from customerip [4500] to myip [4500] (304 bytes)
847358:Jun  3 10:23:34 admin-2 charon: 09[ENC] parsed IKE_AUTH response 1 [ V IDr AUTH SA TSi TSr N(ESP_TFC_PAD_N) N(NON_FIRST_FRAG) N(MOBIKE_SUP) ]
847359:Jun  3 10:23:34 admin-2 charon: 09[IKE] authentication of 'customerip' with pre-shared key successful
847360:Jun  3 10:23:34 admin-2 charon: 09[IKE] IKE_SA vpn-csl[54] established between myip [10.131.112.99]... customerip 
847361:Jun  3 10:23:34 admin-2 charon: 09[IKE] scheduling reauthentication in 85763s
847362:Jun  3 10:23:34 admin-2 charon: 09[IKE] maximum IKE_SA lifetime 86303s
847363:Jun  3 10:23:34 admin-2 charon: 09[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
847364:Jun  3 10:23:34 admin-2 charon: 09[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_512_256/NO_EXT_SEQ
847365:Jun  3 10:23:34 admin-2 charon: 09[IKE] CHILD_SA vpn-test{24} established with SPIs cb2d917d_i 901e6a97_o and TS 192.168.108.40/29 === 10.240.209.128/29

**Reply by tobiasbrunner** [reply]
SA seems to be up, however, the traffic counters do not show any traffic. Also, where is `192.168.108.41` installed if that's the address that's getting pinged? (You only have `192.168.108.40` installed, so you might want to actually install `192.168.108.41/29` on eth0 instead.)

> there is nothing in /etc/strongswan but it is installed, which is odd.

What do you mean?

**Reply by SB8767** [reply]
> SA seems to be up, however, the traffic counters do not show any traffic. Also, where is `192.168.108.41` installed if that's the address that's getting pinged? (You only have `192.168.108.40` installed, so you might want to actually install `192.168.108.41/29` on eth0 instead.)

`>  I thought by giving the eth0:1 interface the subnet mask 255.255.255.248, I will be covered by anything in that subnet.
`
> What do you mean?
`> ignore me. I found the logs.
`

**Reply by tobiasbrunner** [reply]
> I thought by giving the eth0:1 interface the subnet mask 255.255.255.248, I will be covered by anything in that subnet.

No, you have to configure the IP address(es) of your host explicitly. The subnet mask is used to decide if a destination address has to be routed or can be reached directly, not to claim a complete subnet for that host/interface.

**Reply by SB8767** [reply]
ah I see. Thank you for clarifying. So, that means I will have to assign eth0:1 the explicit address of 192.168.108.41. The client asked for a /29 subnet from us, so does that mean I have to create 6 interfaces for each IP address?

**Reply by tobiasbrunner** [reply]
No, why? Do you even need to be reachable on all of these addresses? And even if you are, just add them to eth0. As I mentioned above, the `eth0:x` syntax by `ifconfig` is a legacy feature (these are not actually interfaces, not even virtual ones, these names are simply labels assigned to the configured IP addresses). Should you need any other IP addresses from that subnet, just add them with `ip addr add 192.168.108.4x/29 dev eth0`.

**Reply by SB8767** [reply]
I don't think we do but the customer asked for the subnet. When I run `sudo ip addr`I can see the subnet is saved:

```
 inet 192.168.108.41/29 brd 192.168.108.47 scope global eth0:1
       valid_lft forever preferred_lft forever
```

Is this correct? 

Also, the interface eth0:1 has now been changed to
`eth0:1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.108.41  netmask 255.255.255.248  broadcast 192.168.108.47
        ether 16:a7:4d:68:fa:2f  txqueuelen 1000  (Ethernet)`
        
    The customer still cannot ping 192.168.108.41

**Reply by tobiasbrunner** [reply]
Could be a lot of things, e.g. firewall issues (filtering or NAT), could also be an issue on their end (check the traffic counters for inbound traffic). So you have to debug what happens with the packets.

**Reply by SB8767** [reply]
NAT on my side or theirs? Also, when the tunnel is up and I run tcpdump -i eth0 I don’t see any traffic passing, even with the other vpn I have running. 

If u run tcpdump -i eth0:1 all I see is my home ip accessing the server. 

What is the best way to do a tcpdump for strongswan?

**Reply by tobiasbrunner** [reply]
You should at least see ESP or ESP-in-UDP packets if there is VPN traffic (you'll also see inbound plaintext traffic, but not outbound). As I mentioned previously, you can also check the traffic counters in the status output to see if traffic is exchanged. If there is none while the other end is actually trying to send pings, then the problem might be on their end (but could also be a firewall issue in between e.g. if ESP is blocked, forcing UDP encapsulation might help then).

**Reply by SB8767** [reply]
Thank you. I downloaded iftop and I can't see any traffic coming from the customer public IP.

I have been looking into prerouting and postrouting and I also come across this:
Link: https://docs.strongswan.org/docs/5.9/howtos/forwarding.html

> If the VPN gateway is not the default gateway of the LAN, ICMP redirects might get returned to hosts if they send traffic destined for the remote hosts/subnets to the VPN gateway, directing them to the default gateway of the LAN (which probably doesn’t work and otherwise might get that traffic out unencrypted). To avoid this, disable sending such ICMP messages by setting

> net.ipv4.conf.all.send_redirects=0
> net.ipv4.conf.default.send_redirects=0

Have I read that wrong or does that mean all ICMP traffic is sent back to the host if the VPN gateway is not the default gateway and that could explain why the client cannot ping our eth0:1 IP

**Reply by tobiasbrunner** [reply]
That's only an issue if the pinged host is behind the gateway, i.e. not if an IP of the gateway itself is pinged. As I said before, if you don't see any incoming traffic, the problem is either at the other end or due to a firewall between the two hosts (could be on your end if your firewall rules are restrictive).

**Reply by SB8767** [reply]
Thank you.

I have IPTables that allow my private subnet eth0:1 to the customer private subnet and vice versa.

The VPN is up and I can't see any packets passing. I have a call booked with the customer this Thursday. Can I leave this ticket open until then in case they ask me to make further changes

**Reply by SB8767** [reply]
I just had a call with the customer and I need your advice/ help please.

Customer requested we use the subnet 192.168.108.40/29, so we created a virtual interface of 192.168.108.41 (this was changed following your advice) and the interface was named eth0:1

The LEFTID of the ipsec.conf is my server LAN IP.

The customer said the VPN is going through the tunnel and hitting eth1(LAN IP) first then is forward to 192.168.108.41.

The CX went on to say there should be a DNAT rule to eh0:1, so when traffic goes through the tunnel and hits my side of the VPN, it will hit eth1 then be forwarded to eth0:1, which there is, however, I can't see any traffic the eth1 interface from their source address.

The weird thing is, this ubuntu server already have a VPN up and running without any issues. The customer can ping a different virtual interface and receive replies.

**Reply by tobiasbrunner** [reply]
> Customer requested we use the subnet 192.168.108.40/29, so we created a virtual interface of 192.168.108.41 (this was changed following your advice) and the interface was named eth0:1

Again, that's not an interface, you just added an additional IP address to eth0 (you could also add it to eth1 or lo, or add it directly without the `eth0:1` label, has the same effect).

> The customer said the VPN is going through the tunnel and hitting eth1(LAN IP) first then is forward to 192.168.108.41.

Not sure what they mean or why that would be relevant.

> The CX went on to say there should be a DNAT rule to eh0:1

No, that's not necessary. If the destination address of the tunneled packet is `192.168.108.41`, it's a local address, no matter on what interface it is. There is no need to use DNAT to change the destination address unless they send packets that have a different destination address. However, then this wouldn't work at all as only packets with a destination address in `192.168.108.40/29` even get tunneled (that's the point of IPsec policies).

> however, I can't see any traffic the eth1 interface from their source address.

Which, again, probably means that the problem is on their end (i.e. no packets are sent at all, maybe because they are not matching the negotiated traffic selectors) or between the two hosts (sent packets are dropped by a firewall for some reason). Since you already have a running VPN but you don't see any ESP packets for this new one, it's unlikely the issue is on your end at this point. Unless the existing VPN uses UDP encapsulation (UDP port 4500) but this one doesn't and your firewall drops plain ESP packets (IP protocol 50). So make sure to check your firewall rules.
